### PR TITLE
Thermostat presets

### PR DESCRIPTION
**Pull request by MrMDavidson**
# What does this implement/fix?

This introduces the ability to configure additional presets for a `Thermostat` beyond the existing Home and Away modes. The existing configuration will continue to work. But it also introduces a new configuration list for preset modes. See the example config.

The existing Home/Away logic has been consolidated with this generic preset approach. I'd be keen to further extend this to have the presets also configure the default operation mode/fan mode/swing mode as well as just the temperature.

Apologies for any faux pas in the Python side of it - it's not my forte.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

(Note: Yet to update the docs, wanted to see if this is something of interest to the community/the config stays stable before updating them)

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
climate:
  - platform: thermostat
    # Kept for backwards compatability
    away_config:
      default_target_temperature_low: 16
      default_target_temperature_high: 26

    preset:
      activity:
        default_target_temperature_low: 21
        default_target_temperature_high: 23
      sleep:
        default_target_temperature_low: 17
        default_target_temperature_high: 26
      # etc. Can be any of eco, away, boost, comfort, home, sleep, and activity
    preset_change:
      - logger.log: Preset has been changed
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
This is excellent, and we need this for `bang_bang`, `pid` climate platforms too.

**Comment by kbx81** [comment]
Hey! Thanks for working on this! I think it will make a lot of people happy.

I had a cursory look over the changes and I think it looks good -- I'll try to get around to testing and playing around with it more over the next week or two. 😄

**Comment by MrMDavidson** [comment]
> Hey! Thanks for working on this! I think it will make a lot of people happy.

No worries - it was a bit of a gap in my personal setup so figured I'd have a crack at implementing it. I think the configuration validation will need some tweaks / has some gaps, so happy for any feedback in that area.
 
> I had a cursory look over the changes and I think it looks good -- I'll try to get around to testing and playing around with it more over the next week or two. 😄

Do you think it's worth expanding the functionality over what a preset covers? I'd personally be keen to see the fan mode as part of the preset (Eg. Sleep mode also sets the fan to low) but if I'm doing one it's not much more effort to expand it to others. It does raise questions around if the user changes both the preset and a mode controlled by the preset which takes priority (I'd say the manually specified one?) Eg. Using my example of having the Sleep 
```yaml
  on...:
    then:
      - climate.control:
          id: my_aircon
          preset: sleep
          fan_mode: LOW
```

**Comment by kbx81** [comment]
Haha, well, full disclosure: I don't have strong feelings about presets -- I just don't use them. 😇  I think that, traditionally, presets just define set points, but I don't see anything specifically _wrong_ with allowing the mode/fan mode/swing mode/etc. to be adjusted with them, as well. But that's just my two cents...others here might have different feelings on the matter. 🙃

**Comment by MrMDavidson** [comment]
> Haha, well, full disclosure: I don't have strong feelings about presets -- I just don't use them. 😇 I think that, traditionally, presets just define set points, but I don't see anything specifically _wrong_ with allowing the mode/fan mode/swing mode/etc. to be adjusted with them, as well. But that's just my two cents...others here might have different feelings on the matter. 🙃

Ah, in my case I'm having my controller change presets in response to events. Eg. In the room? Switch to "Activity" preset with a tighter temperature range vs just at home but not in the room vs away from home.

I've actually got another branch with support for controlling these extra modes as part of the presets and it's working well. Will keep playing with it (need to test when they've got a lambda, for instance) and if it works well and there's no opposition I'll merge it into this branch.

**Comment by nagyrobi** [comment]
Can the preset temp values be set at run time? Say, with lambdas pointing at input_numbers?

**Comment by MrMDavidson** [comment]
> Can the preset temp values be set at run time? Say, with lambdas pointing at input_numbers?

I had a quick play but I'm not sure if this is possible overloading the same configuration value for both a compile time value and a run time value. Happy to be corrected, here? But having a look through the docs/code I can't find examples for non-actions that do this. So as far as I can tell this would only be possible if we introduced a second configuration type for when you need it configured via lambda. Something like;
```yaml
    preset:
      activity:
        default_target_temperature_low_templatable: !lambda |-
          return id(my_home_assistant_exported_value).state;
        default_target_temperature_high: 23
```

I think I had a very similar setup to what you're describing and, honestly, it got very messy on the front end and was the driving reason to implement support for presets instead. I can sort of see the use case for lambdas here to return a dynamic value based on something a sensor knows (eg. More than 2 people in the room? Bump the set temperature set points down an extra degree to keep the room cool, or something). That said, these would only be evaluated at switching time (Continuing this example; "There's two people in the room _now_ we're switching to 'Activity' mode" but if another person entered, it wouldn't react to it). With this in mind I think it might be better to leave this functionality out and instead make use of triggers elsewhere in your configuration to prod the state of the A/C? For your use case you can still do something like;

```yaml
  on...
    then:
      - climate.control:
          id: my_aircon
          preset: CLIMATE_PRESET_HOME
          target_temperature_low: !lambda |-
            return id(my_front_end_driven_value_low).state;
          target_temperature_high: !lambda |-
            return id(my_front_end_driven_value_high).state;
```

**Comment by nagyrobi** [comment]
The use case is much more simpler: implementing a control user interface where the various temperatures desired for the presets can be set by the home inhabitants, without needing to recompile node firmware.

So one could configure input numbers (with some defaults, and restore from flash option set) for the various preset modes. The user is free to adjust these as they find it better. For example if you have such a thermostat for each room, it may very well be required that different room inhabitants feel comfortable at different activity or sleep temperatures so they would need to be given the opportunity to customize these values themselves.
Or there may be days when one catches a flu and stays at home, he/she wants higher temps for a few days on their presets...

File: esphome/components/thermostat/climate.py [hunk 1]
```diff
@@ -58,8 +58,14 @@
     CONF_SWING_OFF_ACTION,
     CONF_SWING_VERTICAL_ACTION,
     CONF_TARGET_TEMPERATURE_CHANGE_ACTION,
+    CONF_PRESET,
+    CONF_FAN_MODE,
+    CONF_SWING_MODE,
+    CONF_MODE
 )
 
+CONF_PRESET_CHANGE = "preset_change"
```
Comment (kbx81) [inline esphome/components/thermostat/climate.py:67]: Neat!...but I think if you're going to build a "preset change" trigger, it should be either "on_preset" or "on_preset_change" so it's consistent with other triggers as outlined in the [docs](https://next.esphome.io/guides/automations.html#all-triggers). 😄

File: esphome/components/thermostat/climate.py [hunk 2]
```diff
@@ -183,7 +202,6 @@ def validate_thermostat(config):
                 raise cv.Invalid(
                     f"{req_trigger} must be defined to use {config_trigger}"
                 )
```
Comment (kbx81) [inline esphome/components/thermostat/climate.py]: Total nitpick but...is it really necessary to remove all the blank lines? Visually, for me, they help create separation between blocks of code operating on logically different things. It helps somewhat with readability (I'm a big fan of readability 😄). It also just makes the PR bigger. 😇

**Review by kbx81 — COMMENTED** [review]
One legit concern, one nitpick. Nothing big. 🙂

File: esphome/components/thermostat/climate.py (hunk 2 above)
Comment (MrMDavidson) [reply esphome/components/thermostat/climate.py]: I'm more than happy to leave them - I only stripped them as the build failures suggesting running the code through the Black formatter. The one I happened to be using (online) stripped them out

Comment (kbx81) [reply esphome/components/thermostat/climate.py]: Oh dear -- I hope our linter/formatter configuration isn't stripping them. I'd be really sad. Are you using VSCode? There is an extension so I think you can just do it locally...and I think it'll use our preferred configuration. 🙂

File: esphome/components/thermostat/climate.py (hunk 1 above)
Comment (MrMDavidson) [reply esphome/components/thermostat/climate.py:67]: There's definitely some variance on the naming conventions here but was trying to vaguely make it align with the existing ones for the thermostat (eg. `CONG_AUTO_MODE = "auto_mode"`). My preference would be to make them past tense (`on_preset_changed`) but happy to take your input on preferred naming convention here. Preferences?

File: esphome/components/thermostat/climate.py (hunk 2 above)
Comment (MrMDavidson) [reply esphome/components/thermostat/climate.py]: Alright, let me get that set up and re-lint the files.

File: esphome/components/thermostat/climate.py (hunk 1 above)
Comment (kbx81) [reply esphome/components/thermostat/climate.py:67]: Ah! I'm reading things too fast again. You're adding an action, not a trigger. Not sure what I was thinking...brain went to something else.

I think `preset_change` is fine. Nothing to see here, move along. 😆

**Comment by MrMDavidson** [comment]
Apologies, git rebase gone awry. It's late at night here I'll abandon this and re-open it tomorrow without the mess.

**Comment by kbx81** [comment]
Alrighty! No worries -- if you can fix your branch you can force-push it and it'll fix up the "mess". 😄
If you make a new PR, add a comment and mention this one so we can easily find it. 🙂

**Pull request closed without merging**